crates/engine/tree/src/tree/mod.rs

```diff
@@ -1015,23 +1015,71 @@ where
         version: EngineApiMessageVersion,
     ) -> ProviderResult<TreeOutcome<OnForkChoiceUpdated>> {
         trace!(target: "engine::tree", ?attrs, "invoked forkchoice update");
+
+        // Record metrics
+        self.record_forkchoice_metrics(&attrs);
+
+        // Pre-validation of forkchoice state
+        if let Some(early_result) = self.validate_forkchoice_state(state)? {
+            return Ok(TreeOutcome::new(early_result));
+        }
+
+        // Case 1: Check if head is already canonical
+        if let Some(result) = self.handle_canonical_head(state, attrs.clone(), version)? {
+            return Ok(result);
+        }
+
+        // Case 2 Apply chain update if head is not canonical
+        if let Some(result) = self.apply_chain_update(state, attrs, version)? {
+            return Ok(result);
+        }
+
+        // Case 3 Handle missing block scenario
+        self.handle_missing_block(state)
+    }
```

Comment:
```suggestion
        // Return early if we are on the correct fork
        if let Some(result) = self.handle_canonical_head(state, attrs.clone(), version)? {
            return Ok(result);
        }

        // Attempt to apply a chain update when the head differs from our canonical chain.
	  // This handles reorgs and chain extensions by making the specified head canonical.
        if let Some(result) = self.apply_chain_update(state, attrs, version)? {
            return Ok(result);
        }

	// Fallback that ensures to catch up to the network's state.
        self.handle_missing_block(state)
    }
```
good direction, but instead of saying case, i think it would be better to explain why we return